File: test/fingerprint.test.mjs
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { trackFingerprint, bookFingerprintFrom, fingerprintBook } from '../src/lib/fingerprint.js'

/** Blob determinista de `size` bytes con un patrón reproducible. */
function fakeAudio(size, seed = 1) {
  const bytes = new Uint8Array(size)
  for (let i = 0; i < size; i++) bytes[i] = (i * 31 + seed * 7) % 256
  return new Blob([bytes])
}

test('la misma copia del archivo produce la misma huella', async () => {
  const a = await trackFingerprint(fakeAudio(5000))
  const b = await trackFingerprint(fakeAudio(5000))
  assert.equal(a, b)
  assert.match(a, /^[0-9a-f]{64}$/)
})

test('contenidos distintos producen huellas distintas', async () => {
  const a = await trackFingerprint(fakeAudio(5000, 1))
  const b = await trackFingerprint(fakeAudio(5000, 2))
  assert.notEqual(a, b)
})

test('el tamaño forma parte de la identidad', async () => {
  // Mismo patrón, distinta longitud: no deben colisionar.
  const a = await trackFingerprint(fakeAudio(5000))
  const b = await trackFingerprint(fakeAudio(6000))
  assert.notEqual(a, b)
})

test('archivos mayores de 2 MiB usan primer y ultimo MiB', async () => {
  const big = 3 * 1024 * 1024
  const a = await trackFingerprint(fakeAudio(big))
  const b = await trackFingerprint(fakeAudio(big))
  assert.equal(a, b)
})

test('la huella del libro no depende del orden de las pistas', async () => {
  const t1 = 'a'.repeat(64)
  const t2 = 'b'.repeat(64)
  const t3 = 'c'.repeat(64)
  const ordenado = await bookFingerprintFrom([t1, t2, t3])
  const revuelto = await bookFingerprintFrom([t3, t1, t2])
  assert.equal(ordenado, revuelto)
})

test('fingerprintBook conserva el orden original de las pistas', async () => {
  const blobs = [fakeAudio(1000, 1), fakeAudio(1000, 2)]
  const { bookFingerprint, trackFingerprints } = await fingerprintBook(blobs)
  assert.equal(trackFingerprints.length, 2)
  assert.equal(trackFingerprints[0], await trackFingerprint(blobs[0]))
  assert.equal(trackFingerprints[1], await trackFingerprint(blobs[1]))
  assert.match(bookFingerprint, /^[0-9a-f]{64}$/)
})

/**
 * Vectores congelados. La app de Android DEBE reproducir exactamente estos
 * valores; son la prueba de que ambas implementaciones coinciden.
 * Si un cambio los rompe, el progreso ya sincronizado queda huérfano.
 */
test('vectores de referencia para la implementacion en Kotlin', async () => {
  // Archivo de 10 bytes: 00 01 02 ... 09
  const diezBytes = new Blob([new Uint8Array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])])
  assert.equal(
    await trackFingerprint(diezBytes),
    '83fe3c54f403ec66e809df9dceb0f308fa20394de604b54e9c1a59d805e2e5b7'
  )

  // Libro de dos pistas con huellas conocidas
  assert.equal(
    await bookFingerprintFrom(['00'.repeat(32), 'ff'.repeat(32)]),
    'f7ee6e27721feb087d5ad6f99251059d05183104ae909d2b9830b12cadd4f822'
  )
})

File: android/app/src/test/java/com/lumina/audiolibros/sync/FingerprintTest.kt
package com.lumina.audiolibros.sync

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Test
import java.io.File

/**
 * Verifica que esta implementación coincide con la de la app de escritorio.
 *
 * Los dos primeros valores son los vectores congelados de `docs/SYNC.md`, los
 * mismos que comprueba `test/fingerprint.test.mjs` en el proyecto raíz. Si
 * fallan, los dos dispositivos han dejado de hablar el mismo idioma y la
 * sincronización no emparejará ningún libro.
 */
class FingerprintTest {

    /** Bytes deterministas, iguales que los del test de JavaScript. */
    private fun fakeAudio(size: Int, seed: Int = 1) =
        ByteArray(size) { i -> ((i * 31 + seed * 7) % 256).toByte() }

    @Test
    fun `vector congelado - archivo de diez bytes`() {
        val diezBytes = byteArrayOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
        assertEquals(
            "83fe3c54f403ec66e809df9dceb0f308fa20394de604b54e9c1a59d805e2e5b7",
            Fingerprint.ofBytes(diezBytes)
        )
    }

    @Test
    fun `vector congelado - libro de dos pistas`() {
        assertEquals(
            "f7ee6e27721feb087d5ad6f99251059d05183104ae909d2b9830b12cadd4f822",
            Fingerprint.ofBook(listOf("00".repeat(32), "ff".repeat(32)))
        )
    }

    @Test
    fun `la misma copia produce la misma huella`() {
        assertEquals(Fingerprint.ofBytes(fakeAudio(5000)), Fingerprint.ofBytes(fakeAudio(5000)))
    }

    @Test
    fun `contenidos distintos producen huellas distintas`() {
        assertNotEquals(Fingerprint.ofBytes(fakeAudio(5000, 1)), Fingerprint.ofBytes(fakeAudio(5000, 2)))
    }

    @Test
    fun `el tamano forma parte de la identidad`() {
        assertNotEquals(Fingerprint.ofBytes(fakeAudio(5000)), Fingerprint.ofBytes(fakeAudio(6000)))
    }

    @Test
    fun `archivos mayores de dos MiB usan primer y ultimo MiB`() {
        val grande = 3 * 1024 * 1024
        assertEquals(Fingerprint.ofBytes(fakeAudio(grande)), Fingerprint.ofBytes(fakeAudio(grande)))
    }

    @Test
    fun `el orden de las pistas no altera la huella del libro`() {
        val a = "a".repeat(64)
        val b = "b".repeat(64)
        val c = "c".repeat(64)
        assertEquals(Fingerprint.ofBook(listOf(a, b, c)), Fingerprint.ofBook(listOf(c, a, b)))
    }

    @Test
    fun `leer de un archivo da el mismo resultado que leer de memoria`() {
        val contenido = fakeAudio(3 * 1024 * 1024 + 517)
        val temporal = File.createTempFile("huella", ".bin")
        try {
            temporal.writeBytes(contenido)
            assertEquals(Fingerprint.ofBytes(contenido), Fingerprint.ofTrack(temporal))
        } finally {
            temporal.delete()
        }
    }
}
